File: 3주차/서재은/주차요금계산.py
import math
def time_calculate(i, o):
    # H 계산
    h = (int(o.split(':')[0]) - int(i.split(':')[0])) * 60
    # M 계산
    m = int(o.split(':')[1]) - int(i.split(':')[1])
    return h + m

def solution(fees, records):
    in_time = {}
    parked_acc = {}
    for log in records:
        time = log.split()[0]; number = log.split()[1]; status = log.split()[2]
        if number in in_time.keys():
            # in_time[number], time의 차 계산해서 parked_acc에 추가
            if number not in parked_acc.keys():
                parked_acc[number] = time_calculate(in_time[number], time)
            else:
                parked_acc[number] += time_calculate(in_time[number], time)
            in_time.pop(number)
        elif number not in in_time.keys():
            in_time[number] = time
    # 출차하지 않은 차들의 요금을 계산한다.
    for number,time in in_time.items():
        if number not in parked_acc.keys():
            parked_acc[number] = time_calculate(in_time[number], '23:59')
        else:
            parked_acc[number] += time_calculate(in_time[number], '23:59')

    answer = []
    for v in dict(sorted(parked_acc.items())).values():
        print(v)
        if int(v)>fees[0]:
            answer.append(fees[1] + math.ceil((int(v)-fees[0])/fees[2]) * fees[3])
        else:
            answer.append(fees[1])
    return answer

File: 3주차/이동준/주차요금계산.java
import java.util.*;

class Solution {
    public static int[] solution(int[] fees, String[] records) {
        Map<String, Integer> park = new HashMap<>();  // 차번호 ,시간8
        // 차번호, 요금 / 차번호 기준 오름차순 정렬
        Map<String, Integer> cost = new TreeMap<>();

        for (String r : records) {
            String[] record = r.split(" ");
            String[] t = record[0].split(":");
            int time = Integer.parseInt(t[0]) * 60 + Integer.parseInt(t[1]);
            String num = record[1];
            String in_out = record[2];

            if (in_out.equals("IN")) {
                park.put(num, time);
            } else if (in_out.equals("OUT")) {
                if (!cost.containsKey(num)) { // 처음 들어온 경우
                    cost.put(num, time - park.get(num));
                } else {
                    cost.put(num, cost.get(num) + time - park.get(num));
                }
                park.remove(num);
            }
        }

        // 아직 나가지 않은 차량
        if (!park.isEmpty()) {
            for (String carNum : park.keySet()) {
                int accumulatedTime = cost.getOrDefault(carNum, 0);
                cost.put(carNum, accumulatedTime + (23 * 60 + 59) - park.get(carNum));
            }
        }

        List<Integer> answer = new ArrayList<>(cost.size());
        for (Integer c : cost.values()) {
            int baseTime = fees[0];
            int baseFee = fees[1];
            int unitTime = fees[2];
            int unitFee = fees[3];

            // 요금 = 기본 요금 + ((총 주차 시간 - 기본 시간) / 단위 시간) * 단위 요금
            if (c <= baseTime) {
                answer.add(baseFee);
            } else {
                answer.add(baseFee + (int)Math.ceil((double)(c - baseTime) / unitTime) * unitFee);
            }
        }
        return answer.stream().mapToInt(Integer::intValue).toArray();
    }
}
